File: packages/ts/spaces/src/node-keyring.ts
/**
 * Per-node keyring helpers — E2EE primitive for invite-node encryption.
 *
 * Each `invite+enc` node carries its OWN keyring whose CEK is wrapped ONLY to
 * that node's participants, not the space-wide keyring. An external requester
 * can therefore read/write their content E2EE without holding the space key.
 *
 * These are thin wrappers over the generic keyring helpers in `client.ts`,
 * specialised to the node keyring path so call sites cannot accidentally target
 * the space keyring.
 *
 * INVARIANT: `ownerEnsureNodeKeyring` MUST run before `addNodeKeyringRecipient`.
 * Use `ensureNodeKeyringRecipient` to get both in the correct order.
 */
import { removeRecipient } from "@drakkar.software/starfish-keyring"
import type { Encryptor, StarfishClient } from "@drakkar.software/starfish-client"

import { openEncryptor, buildEncryptor, ownerEnsureKeyring, addKeyringRecipientCore } from "./client.js"
import type { DeviceKeys } from "./client.js"
import { ownerTrustedAdders } from "./session.js"
import type { Session } from "./session.js"
import { computeOwnerTrustedAdders } from "@drakkar.software/starfish-identities"

export { computeOwnerTrustedAdders }

/** A keyring recipient referenced by their X25519 KEM pubkey (hex). */
export interface NodeKeyringRecipient {
  subKem: string
  userId?: string
  label?: string
}

/**
 * Owner/creator side: create the node keyring if missing, then return an encryptor.
 */
export function ownerEnsureNodeKeyring(
  session: Session,
  spaceId: string,
  nodeId: string,
  trustedAdders: string[] = ownerTrustedAdders(session),
): Promise<Encryptor> {
  return ownerEnsureKeyring(
    session.contentClient,
    session.keys,
    session.layout.nodeKeyringPull(spaceId, nodeId),
    session.layout.nodeKeyringPush(spaceId, nodeId),
    trustedAdders,
  )
}

/**
 * Open the node keyring as a recipient (throws `SpaceAccessError` if absent
 * or if the caller is not a recipient).
 */
export function openNodeEncryptor(
  client: StarfishClient,
  keys: DeviceKeys,
  session: Pick<Session, "layout">,
  spaceId: string,
  nodeId: string,
  trustedAdders: string[],
): Promise<Encryptor> {
  return openEncryptor(client, keys, session.layout.nodeKeyringPull(spaceId, nodeId), trustedAdders)
}

/** Soft variant of {@link openNodeEncryptor}: resolves `null` instead of throwing.
 *
 * Propagates `NodeAccessRevokedError` when the server returns 403 — a revocation
 * signal that callers must handle explicitly rather than treating as "not ready".
 */
export function buildNodeEncryptor(
  client: StarfishClient,
  keys: DeviceKeys,
  session: Pick<Session, "layout">,
  spaceId: string,
  nodeId: string,
  trustedAdders: string[],
): Promise<Encryptor | null> {
  return buildEncryptor(client, keys, session.layout.nodeKeyringPull(spaceId, nodeId), trustedAdders, spaceId, nodeId)
}

/**
 * Add a recipient to the node keyring. The keyring MUST already exist (call
 * {@link ownerEnsureNodeKeyring} first). "Already present" is swallowed so
 * re-inviting the same KEM is idempotent.
 */
export function addNodeKeyringRecipient(
  session: Session,
  spaceId: string,
  nodeId: string,
  recipient: NodeKeyringRecipient,
  opts: { trustedAdders?: string[] } = {},
): Promise<void> {
  return addKeyringRecipientCore(
    session.contentClient,
    session.keys,
    session.layout.nodeKeyringName(spaceId, nodeId),
    recipient,
    opts.trustedAdders ?? ownerTrustedAdders(session),
  )
}

/**
 * Ensure the node keyring exists, then add a recipient — in that order.
 * Returns the owner's encryptor so the creator can immediately seal.
 */
export async function ensureNodeKeyringRecipient(
  session: Session,
  spaceId: string,
  nodeId: string,
  recipient: NodeKeyringRecipient,
  opts: { trustedAdders?: string[] } = {},
): Promise<Encryptor> {
  const enc = await ownerEnsureNodeKeyring(session, spaceId, nodeId, opts.trustedAdders)
  await addNodeKeyringRecipient(session, spaceId, nodeId, recipient, opts)
  return enc
}

/** Who to use as the keyring adder when revoking. Defaults to the session's own edPub. */
function adderOf(session: Session) {
  return { edPriv: session.keys.edPriv, edPub: session.keys.edPub, kemPriv: session.keys.kemPriv }
}

/**
 * REVOKE recipients from a node keyring: rotates to a NEW epoch, mints a fresh
 * CEK, and re-wraps ONLY to retained recipients. Removed parties lose access to
 * future messages (forward secrecy only). Returns the new epoch number.
 */
export async function removeNodeKeyringRecipient(
  session: Session,
  spaceId: string,
  nodeId: string,
  removeSubKems: string[],
  opts: { trustedAdders?: string[] } = {},
): Promise<{ newEpoch: number }> {
  return removeRecipient(
    session.contentClient,
    session.layout.nodeKeyringName(spaceId, nodeId),
    removeSubKems,
    adderOf(session),
    { trustedAdders: opts.trustedAdders ?? ownerTrustedAdders(session) },
  )
}

File: packages/python/spaces/starfish_spaces/node_keyring.py
"""Per-node keyring helpers (E2EE for invite-node encryption).

Thin wrappers over :mod:`starfish_spaces.client` keyring helpers specialised to
the node-keyring path.  INVARIANT: ensure the keyring exists before adding a
recipient.
"""

from __future__ import annotations

from typing import TYPE_CHECKING, Any, Optional, TypedDict

from starfish_keyring import remove_recipient
from starfish_identities import compute_owner_trusted_adders

from starfish_spaces.client import (
    DeviceKeys,
    add_keyring_recipient_core,
    build_encryptor,
    open_encryptor,
    owner_ensure_keyring,
)

if TYPE_CHECKING:
    from starfish_sdk import StarfishClient
    from starfish_spaces.config import SpaceLayout
    from starfish_spaces.session import Session


# ── Types ─────────────────────────────────────────────────────────────────────


class NodeKeyringRecipient(TypedDict, total=False):
    """A recipient to add/remove from a per-node keyring."""

    subKem: str
    userId: Optional[str]
    label: Optional[str]


# ── Internal helper ───────────────────────────────────────────────────────────


def _adder_of(session: "Session") -> dict[str, str]:
    return {
        "edPriv": session.keys["edPriv"],
        "edPub": session.keys["edPub"],
        "kemPriv": session.keys["kemPriv"],
    }


# ── Public API ────────────────────────────────────────────────────────────────


async def owner_ensure_node_keyring(
    client: "StarfishClient",
    session: "Session",
    space_id: str,
    node_id: str,
    layout: "SpaceLayout",
) -> None:
    """Ensure the per-node keyring exists and the owner is in it.

    CAS-safe (retries on conflict via :func:`owner_ensure_keyring`).
    """
    trusted = owner_trusted_adders(session)
    await owner_ensure_keyring(
        client,
        session.keys,  # type: ignore[arg-type]
        layout.node_keyring_name(space_id, node_id),
        layout.node_keyring_pull(space_id, node_id),
        layout.node_keyring_push(space_id, node_id),
        trusted_adders=trusted,
    )


async def open_node_encryptor(
    client: "StarfishClient",
    session: "Session",
    space_id: str,
    node_id: str,
    layout: "SpaceLayout",
) -> Any:
    """Open an encryptor for the per-node keyring (raises :class:`SpaceAccessError` on failure)."""
    trusted = owner_trusted_adders(session)
    return await open_encryptor(
        client,
        layout.node_keyring_name(space_id, node_id),
        session.keys["kemPriv"],
        trusted_adders=trusted,
        space_id=space_id,
        node_id=node_id,
    )


async def build_node_encryptor(
    client: "StarfishClient",
    session: "Session",
    space_id: str,
    node_id: str,
    layout: "SpaceLayout",
) -> Optional[Any]:
    """Like :func:`open_node_encryptor` but returns ``None`` instead of raising."""
    trusted = owner_trusted_adders(session)
    return await build_encryptor(
        client,
        layout.node_keyring_name(space_id, node_id),
        session.keys["kemPriv"],
        trusted_adders=trusted,
        space_id=space_id,
        node_id=node_id,
    )


async def add_node_keyring_recipient(
    client: "StarfishClient",
    session: "Session",
    space_id: str,
    node_id: str,
    recipient: NodeKeyringRecipient,
    layout: "SpaceLayout",
) -> None:
    """Add ``recipient`` to the per-node keyring."""
    trusted = owner_trusted_adders(session)
    await add_keyring_recipient_core(
        client,
        layout.node_keyring_name(space_id, node_id),
        dict(recipient),
        _adder_of(session),
        trusted_adders=trusted,
    )


async def ensure_node_keyring_recipient(
    client: "StarfishClient",
    session: "Session",
    space_id: str,
    node_id: str,
    recipient: NodeKeyringRecipient,
    layout: "SpaceLayout",
) -> None:
    """Ensure the node keyring exists, then add ``recipient``."""
    await owner_ensure_node_keyring(client, session, space_id, node_id, layout)
    await add_node_keyring_recipient(client, session, space_id, node_id, recipient, layout)


async def remove_node_keyring_recipient(
    client: "StarfishClient",
    session: "Session",
    space_id: str,
    node_id: str,
    sub_kem: str,
    layout: "SpaceLayout",
) -> None:
    """Remove ``sub_kem`` from the per-node keyring (forward secrecy via epoch rotation).

    Removes and rotates the keyring epoch so the removed recipient cannot decrypt
    future content even if they retained the old CEK.
    """
    trusted = owner_trusted_adders(session)
    await remove_recipient(
        client,
        layout.node_keyring_name(space_id, node_id),
        [sub_kem],
        _adder_of(session),  # type: ignore[arg-type]
        trusted_adders=trusted,
    )


def owner_trusted_adders(session: "Session") -> list[str]:
    """Return the trusted-adder allow-list for a session opening its own keyring."""
    return compute_owner_trusted_adders(session.owner_ed_pub, session.keys["edPub"])


__all__ = [
    "NodeKeyringRecipient",
    "owner_ensure_node_keyring",
    "open_node_encryptor",
    "build_node_encryptor",
    "add_node_keyring_recipient",
    "ensure_node_keyring_recipient",
    "remove_node_keyring_recipient",
    "owner_trusted_adders",
    "compute_owner_trusted_adders",
]
